### PR TITLE
Fixes

### PR DESCRIPTION
**Pull request by likholat**

### CODE REVIEW

File: plan.py [hunk 1]
```diff
@@ -100,38 +112,41 @@ def rollout(node, max_depth=200):
         # Select a random mol (that's not a starting mol)
         mols = [mol for mol in cur.state if mol not in starting_mols]
         mol = random.choice(mols)
-        fprs = policies.fingerprint_mols([mol])
+        fprs = policies.fingerprint_mols([mol], fp_dim=8912)
```
Comment (dkurt) [inline plan.py]: Can we take 8912 from input? Like a shape of input data.

File: policies.py [hunk 2]
```diff
@@ -86,98 +80,104 @@ def train(sess, net, X, y, batch_size=16, epochs=10):
...
 
 class RolloutPolicyNet:
     def __init__(self, n_rules, fp_dim=8912, k=10):
         self.fp_dim = fp_dim
         self.n_rules = n_rules
-        self.X = tf.placeholder(tf.float32, shape=(None, fp_dim))
-        self.y = tf.placeholder(tf.int64, shape=(None,))
-        self.keep_prob = tf.placeholder(tf.float32, shape=())
+        self.X = tf.placeholder(tf.float32, shape=(None, fp_dim), name='X')
+        self.y = tf.placeholder(tf.int64, shape=(None,), name='y')
```
Comment (dkurt) [inline policies.py:106]: Can we use default names?

**Comment by dkurt** [comment]
Please align fork's master with origin

File: plan.py [hunk 3]
```diff
@@ -8,6 +8,14 @@
 from rdkit import Chem
 from rdkit.Chem import AllChem
 
+import numpy as np
+import argparse
+from openvino.inference_engine import IECore
+
+parser = argparse.ArgumentParser("Select model representation")
+parser.add_argument("-use_openvino", help="Convert Tensorflow model to OpenVino IR model.", action='store_true')
```
Comment (dkurt) [inline plan.py]: `--use_openvino`

File: .github/workflows/main.yml [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        pip install test-generator
+        conda install -c rdkit rdkit=2018.09.1
+        conda install -c conda-forge molvs
+        conda install -c anaconda networkx
+        conda install -c anaconda defusedxml
+
+    - name: Run OpenVINO model
+      run: |
+        source /opt/intel/openvino/bin/setupvars.sh
+        export PYTHONPATH=/opt/intel/openvino/deployment_tools/model_optimizer:$PYTHONPATH
```
Comment (dkurt) [inline .github/workflows/main.yml]: Do we need this step?

File: plan.py [hunk 5]
```diff
@@ -8,13 +8,26 @@
 from rdkit import Chem
 from rdkit.Chem import AllChem
 
+import numpy as np
+import argparse
+
+parser = argparse.ArgumentParser("Select model representation")
+parser.add_argument("--use_openvino", help="Convert Tensorflow model to OpenVINO IR model.", action='store_true')
```
Comment (dkurt) [inline plan.py]: Change to more intuitive description - "Enable OpenVINO optimization"

File: plan.py [hunk 6]
```diff
@@ -33,6 +46,46 @@
...
 
+if args.use_openvino:
+    import subprocess
+    import sys
+    import mo_tf
+
+    from openvino.inference_engine import IECore
+
+    # Save frozen graph
+    pb_model_path = 'model/model.pb'
```
Comment (dkurt) [inline plan.py]: Let's save just `model.pb` and then remove programmatically after conversion.

File: plan.py [hunk 7]
```diff
@@ -33,6 +46,46 @@
...
+
+    with tf.io.gfile.GFile(pb_model_path, 'wb') as f:
+        f.write(graph_def.SerializeToString())
+
+    # Convert to OpenVINO IR
+    subprocess.run(
+        [
+            sys.executable, mo_tf.__file__, '--input_model', pb_model_path,
+            '--output_dir', 'model/', '--input', 'Placeholder,Placeholder_2', '--input_shape', "[1, 10000],[1]",
+            '--freeze_placeholder_with_value', "Placeholder_2->5"
```
Comment (dkurt) [inline plan.py]: 5?

File: plan.py [hunk 8]
```diff
@@ -55,39 +108,85 @@ def expansion(node):
...
-        expansion_net.keep_prob: 1.,
-        expansion_net.X: fprs,
-        expansion_net.k: 5
-    })
+        # out_node = sess.graph.get_tensor_by_name('Softmax_1:0')
+        # preds = sess.run(out_node, feed_dict={
+        #     expansion_net.keep_prob: 1.,
+        #     expansion_net.X: fprs,
+        #     expansion_net.k: 5
+        # })
```
Comment (dkurt) [inline plan.py]: please remove commented code

File: plan.py (hunk 7 above)
Comment (likholat) [reply plan.py]: We should get top 5 predictions

Comment (dkurt) [reply plan.py:78]: Can we avoid this hardcoded number? Perform TopK in python (so output of network will be Softmax_1) or pass by input?

File: plan.py [hunk 9]
```diff
@@ -33,6 +46,48 @@
...
+            sys.executable, mo_tf.__file__, '--input_model', pb_model_path,
+            '--output_dir', 'model/', '--input', 'Placeholder,Placeholder_2', '--input_shape', "[1, 10000],[1]",
+            '--freeze_placeholder_with_value', "Placeholder_2->5"
+        ],
+        check=True)
+
+    os.remove(pb_model_path)
+
+    model_xml = 'model/model.xml'
+    model_bin = 'model/model.bin'
```
Comment (dkurt) [inline plan.py]: same for .xml and .bin - Let's remove after loading and use current directory

File: plan.py [hunk 10]
```diff
@@ -55,39 +110,78 @@ def expansion(node):
...
+        np.save('expansion_preds_tf.npy', [indices, values])
 
+    else:
+        # Get output nodes names
+        input_blob = next(iter(net.inputs))
+        values_blob = list(iter(net.outputs))[0]
+        indices_blob = list(iter(net.outputs))[1]
+
+        # Predict applicable rules
+        preds = exec_net.infer(inputs={input_blob: fprs})
```
Comment (dkurt) [inline plan.py]: Can we use `expansion_net.X` to get input name?

File: policies.py [hunk 11]
```diff
@@ -104,10 +104,12 @@ def __init__(self, n_rules, fp_dim=8912, k=10):
         self.n_rules = n_rules
         self.X = tf.placeholder(tf.float32, shape=(None, fp_dim), name='X')
         self.y = tf.placeholder(tf.int64, shape=(None,), name='y')
+        self.keep_prob = keep_prob
 
-        inp = tf.math.log(self.X+1)
+        # inp = tf.math.log(self.X+1)
+        inp = self.X+1
```
Comment (dkurt) [inline policies.py]: Please check what should be done in MO to enable Log. Open a pull request to OpenVINO repo and use this branch in the tests to convert a model.

File: plan.py (hunk 10 above)
Comment (likholat) [reply plan.py]: `expansion_net.X.name` is `Placeholder:0`, and `input_blob` must be just `Placeholder`, so I should write: 
```
input_blob = expansion_net.X.name.split(':')[0]
```
Is it better?

Comment (dkurt) [reply plan.py]: Yes, that's fine!

File: .github/workflows/main.yml [hunk 12]
```diff
@@ -32,7 +31,11 @@ jobs:
         tar -xf model/model.tar.gz -C model
 
     - name: Check opvn
-      run: find -name "openvino"
+      run: |
+        git clone https://github.com/likholat/openvino.git && cd openvino
+        git fetch origin tf_log_op
```
Comment (dkurt) [inline .github/workflows/main.yml]: merged

File: .github/workflows/main.yml [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+        conda install -c conda-forge molvs
+        conda install -c anaconda networkx
+        conda install -c anaconda defusedxml
+
+    - name: Run OpenVINO model
+      env:
+        ACTIONS_ALLOW_UNSECURE_COMMANDS: 'true'
+      run: |
+        find -name "mo_tf.py"
+        echo "::set-env name=PYTHONPATH::./openvino/model-optimizer" >> $GITHUB_ENV
```
Comment (dkurt) [inline .github/workflows/main.yml]: incorrect usage. Try
```
source /opt/intel/openvino/bin/setupvars.sh
export PYTHONPATH=./openvino/model-optimizer:$PYTHONPATH
```
export should go **after** source

File: .github/workflows/main.yml [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+        sudo apt-get update
+        sudo apt-get install -y --no-install-recommends intel-openvino-dev-ubuntu18-2021.1.110
+        sudo ln -s /opt/intel/openvino_2021 /opt/intel/openvino
+
+    - name: Install dependencies
+      run: |
+        conda install python=3.6
+        pip install tensorflow==1.15.2
+        pip install test-generator
+        pip install onnx
```
Comment (dkurt) [inline .github/workflows/main.yml]: No longer needed

File: plan.py [hunk 15]
```diff
@@ -137,7 +137,7 @@ def transform(mol, rule):
     results = rxn.RunReactants([mol])
 
     # Only look at first set of results (TODO any reason not to?)
-    # results = results[0]
+    results = results[0] if results else results
```
Comment (dkurt) [inline plan.py]: better to make it more intuitive:
```python
if not results:
  return []
```

File: policies.py [hunk 16]
```diff
@@ -104,10 +104,11 @@ def __init__(self, n_rules, fp_dim=8912, k=10):
         self.n_rules = n_rules
         self.X = tf.placeholder(tf.float32, shape=(None, fp_dim), name='X')
         self.y = tf.placeholder(tf.int64, shape=(None,), name='y')
+        self.keep_prob = keep_prob
 
         inp = tf.math.log(self.X+1)
         net = tf.layers.dense(inp, 512, activation=tf.nn.elu)
-        net = tf.nn.dropout(net, keep_prob=keep_prob)
+        # net = tf.nn.dropout(net, keep_prob=keep_prob)
```
Comment (dkurt) [inline policies.py]: add `is_training` flag to this class to build with or without dropout layers

File: plan.py [hunk 17]
```diff
@@ -55,26 +157,48 @@ def expansion(node):
...
-        new_state = mols - {mol}
+
+        smol = set(mol)
+        new_state = [x for x in mols if x in smol]
 
         mol = Chem.MolFromSmiles(mol)
         for idx in rule_idxs:
             # Extract actual rule
-            rule = expansion_rules[idx]
+            rule = list(expansion_rules.keys())[list(expansion_rules.values()).index(idx)]
```
Comment (dkurt) [inline plan.py]: Hard to understand this fix. Can you please comment?

File: plan.py [hunk 18]
```diff
@@ -55,26 +157,48 @@ def expansion(node):
     # Convert mols to format for prediction
     # If the mol is in the starting set, ignore
     mols = [mol for mol in mols if mol not in starting_mols]
-    fprs = policies.fingerprint_mols(mols)
+    fprs = policies.fingerprint_mols(mols, expansion_net.X.shape[1])
+
+    if not args.use_openvino:
+        # Predict applicable rules
+        preds = sess.run(expansion_net.pred, feed_dict={
+            expansion_net.keep_prob: 1.,
```
Comment (dkurt) [inline plan.py]: Can we skip this placeholder now when we introduced `is_training`?

File: plan.py [hunk 19]
```diff
@@ -33,12 +47,100 @@
...
+    # Convert to OpenVINO IR
+    subprocess.run(
+        [
+            sys.executable, mo_tf.__file__, '--input_model', pb_model_path,
+            '--input', ','.join(input_node_names), '--input_shape', "[1, 8912]",
+        ],
+        check=True)
+
+    roll_net = ie.read_network(model=model_xml, weights=model_bin)
+    roll_exec_net = ie.load_network(network=roll_net, device_name='CPU')
```
Comment (dkurt) [inline plan.py]: Named arguments is a good practice but in this case variables names are meaningful enough. So let's avoid named arguments in these two lines.

File: plan.py [hunk 20]
```diff
@@ -33,12 +47,100 @@
...
+            graph_def, input_node_names, output_node_names, tf.float32.as_datatype_enum)
+
+    with tf.io.gfile.GFile(pb_model_path, 'wb') as f:
+        f.write(graph_def.SerializeToString())
+
+    # Convert to OpenVINO IR
+    subprocess.run(
+        [
+            sys.executable, mo_tf.__file__, '--input_model', pb_model_path,
+            '--input', ','.join(input_node_names), '--input_shape', "[1, 8912]",
```
Comment (dkurt) [inline plan.py]: Can we remove `'--input', ','.join(input_node_names),` flag?

File: plan.py [hunk 21]
```diff
@@ -33,12 +47,100 @@
...
+    output_node_names = ['TopKV2']
+
+    graph_def = tf.graph_util.convert_variables_to_constants(
+            sess,
+            sess.graph_def,
+            output_node_names)
+
+    from tensorflow.python.tools import optimize_for_inference_lib
+    graph_def = optimize_for_inference_lib.optimize_for_inference(
+            graph_def, input_node_names, output_node_names, tf.float32.as_datatype_enum)
```
Comment (dkurt) [inline plan.py]: Will code work without `optimize_for_inference` step?

File: plan.py [hunk 22]
```diff
@@ -100,16 +224,30 @@ def rollout(node, max_depth=200):
...
+            # Get output node name
+            output_blob = next(iter(roll_net.outputs))
+
+            # Predict applicable rules
+            preds = roll_exec_net.infer(inputs={roll_inp_node_name: fprs})
+            preds = preds[output_blob]
+
+            np.save('rollout_preds_ov', preds[0])
+
+            print('HELLOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO')
```
Comment (dkurt) [inline plan.py]: ?

File: plan.py [hunk 23]
```diff
@@ -100,16 +224,30 @@ def rollout(node, max_depth=200):
...
+
+            np.save('rollout_preds_tf', preds[0])
+
+        else:
+            # Get output node name
+            output_blob = next(iter(roll_net.outputs))
+
+            # Predict applicable rules
+            preds = roll_exec_net.infer(inputs={roll_inp_node_name: fprs})
+            preds = preds[output_blob]
```
Comment (dkurt) [inline plan.py]: can be replaced to `preds = next(iter(preds.values()))`?

File: plan.py (hunk 17 above)
Comment (likholat) [reply plan.py]: `rule` is a `dict`, where idx is value and rule is key, so we should get the key by value.

File: plan.py (hunk 21 above)
Comment (likholat) [reply plan.py]: Yes, should I remove it?

File: plan.py (hunk 20 above)
Comment (likholat) [reply plan.py]: Yes

File: plan.py (hunk 21 above)
Comment (dkurt) [reply plan.py]: Yes, if we can work without it

File: plan.py (hunk 17 above)
Comment (dkurt) [reply plan.py]: Can you please show an example? Value of `expansion_rules` and one of the `idx` + `rule`

Comment (likholat) [reply plan.py]: `expansion_rules` is too large to show, `rule` + `idx`:  
```
('[C&H3:12]-[C:11]1(-[#7:10]2-[C:8](=[O:9])-[c:7]3:[c:2](-[N&H2:1]):[c:3]:[c:4]:[c:5]:[c:6]:3-[C:20]-2=[O:21])-[C:13]-[C:14]-[C:15](=[O:16])-[#7:17]-[C:18]-1=[O:19]>>[#7;+:1]-[c:2]1:[c:3]:[c:4]:[c:5]:[c:6]2:[c:7]:1-[C:8](=[O:9])-[#7:10](-[C:11]1(-[C&H3:12])-[C:13]-[C:14]-[C:15](=[O:16])-[#7:17]-[C:18]-1=[O:19])-[C:20]-2=[O:21]', 0)
```

Comment (dkurt) [reply plan.py]: Can we just swap keys and values this way?
```python
expansion_rules = rules['expansion']
expansion_rules = {idx: rule for rule, idx in expansion_rules.items()}
```

File: plan.py [hunk 24]
```diff
@@ -8,23 +8,38 @@
 from rdkit import Chem
 from rdkit.Chem import AllChem
 
+import numpy as np
+import argparse
+from tensorflow.python.ops import gen_nn_ops
```
Comment (dkurt) [inline plan.py]: unused import

File: plan.py [hunk 25]
```diff
@@ -55,21 +153,46 @@ def expansion(node):
...
-    })
+    if not args.use_openvino:
+        # Predict applicable rules
+        preds = sess.run(expansion_net.pred, feed_dict={
+            expansion_net.X: fprs,
+            expansion_net.k: 5
+        })
+
+        # indices = preds.indices[0]
+        # values = preds.values[0]
```
Comment (dkurt) [inline plan.py]: remove commented code

File: plan.py [hunk 26]
```diff
@@ -152,9 +212,88 @@ def plan(target_mol):
...
+        exp_inp_node_x_name = expansion_net.X.name.split(':')[0]
+        exp_inp_node_k_name = expansion_net.k.name.split(':')[0]
+        roll_inp_node_name = rollout_net.X.name.split(':')[0]
+
+        exp_input_nodes = [exp_inp_node_x_name, exp_inp_node_k_name]
+        exp_output_nodes = ['TopKV2']
+        roll_output_nodes = ['ArgMax']
+
+        exp_net, exp_exec_net = convert_to_ir(sess, [exp_net_x_shape, 0], exp_output_nodes, exp_input_nodes, "Placeholder_2->5")
+        _, roll_exec_net = convert_to_ir(sess, [roll_net_x_shape], roll_output_nodes)
```
Comment (dkurt) [inline plan.py:283]: Seems to me that this code block should be inside `plan` method

File: mcts.py [hunk 27]
```diff
@@ -56,13 +56,14 @@ def mcts(root, expansion_policy, rollout_policy, iterations=2000, max_depth=200)
 
             # Expansion
             s = time()
-            cur_node.children = expansion_policy(cur_node)
+            cur_node.children = expansion_policy(cur_node, use_openvino=use_openvino)
             print('Expansion took:', time() - s)
-            cur_node = cur_node.best_child()
+            if cur_node.children:
+                cur_node = cur_node.best_child()
```
Comment (dkurt) [inline mcts.py:62]: We we cannot run this method? If not, `cur_node` is uninitialized.
